### PR TITLE
Simplify Snapshot Initialization

### PR DESCRIPTION
**Pull request by original-brownbear**
We were loading `RepositoryData` twice during snapshot initialization,
redundantly checking if a snapshot existed already.
The first snapshot existence check is somewhat redundant because a snapshot could be
created between loading `RepositoryData` and updating the cluster state with the `INIT`
state snapshot entry.
Also, it is much safer to do the subsequent checks for index existence in the repo and
and the presence of old version snapshots once the `INIT` state entry prevents further
snapshots from being created concurrently.
While the current state of things will never lead to corruption on a concurrent snapshot
creation, it could result in a situation (though unlikely) where all the snapshot's work
is done on the data nodes, only to find out that the repository generation was off during
snapshot finalization, failing there and leaving a bunch of dead data in the repository
that won't be used in a subsequent snapshot (because the shard generation was never referenced
due to the failed snapshot finalization).
BwC should not be a concern here since the init stage only has meaning for a single master node as any init stage snapshot is removed on master-failover so creating the placeholder entry with repo generation `-2` makes no difference here..

Note: This is a step on the way to parallel repository operations by making snapshot related CS
and repo related CS more tightly correlated.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -268,90 +268,86 @@ public void createSnapshot(final CreateSnapshotRequest request, final ActionList
...
+                    throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName,
+                        "cannot snapshot while a snapshot deletion is in-progress in [" + deletionsInProgress + "]");
                 }
+                final RepositoryCleanupInProgress repositoryCleanupInProgress = currentState.custom(RepositoryCleanupInProgress.TYPE);
+                if (repositoryCleanupInProgress != null && repositoryCleanupInProgress.hasCleanupInProgress()) {
+                    throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName,
+                        "cannot snapshot while a repository cleanup is in-progress in [" + repositoryCleanupInProgress + "]");
+                }
+                SnapshotsInProgress snapshots = currentState.custom(SnapshotsInProgress.TYPE);
+                if (snapshots == null || snapshots.entries().isEmpty()) {
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: I find this a bit more readable:

```suggestion
                if (snapshots != null && snapshots.entries().isEmpty() == false) {
                        throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName, " a snapshot is already running");
                }
                // Store newSnapshot here to be processed in clusterStateProcessed
                ...
```

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 2]
```diff
@@ -268,90 +268,86 @@ public void createSnapshot(final CreateSnapshotRequest request, final ActionList
...
+                    // Store newSnapshot here to be processed in clusterStateProcessed
+                    indices = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(currentState,
+                        request.indicesOptions(), request.indices()));
+                    logger.trace("[{}][{}] creating snapshot for indices [{}]", repositoryName, snapshotName, indices);
+                    newSnapshot = new SnapshotsInProgress.Entry(
+                        new Snapshot(repositoryName, snapshotId),
+                        request.includeGlobalState(), request.partial(),
+                        State.INIT,
+                        Collections.emptyList(),
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Maybe add a comment on why the list is empty here? Something like 
// list of snapshot indices will be resolved later

**Review by tlrx — APPROVED** [review]
LGTM, makes sense. I only left very minor comments.

**Comment by original-brownbear** [comment]
Thanks Tanguy!

**Pull request merged**